### PR TITLE
Implement `mutate/2` as a macro using `Explorer.Query`

### PR DESCRIPTION
**Pull request by philss**
This is part of https://github.com/elixir-nx/explorer/issues/402

This also refactors the filtering tests to use the macro, and includes more binary operators.

### CODE REVIEW

File: lib/explorer/data_frame.ex [hunk 1]
```diff
@@ -1329,8 +1327,11 @@ defmodule Explorer.DataFrame do
   Picks rows based on `Explorer.Query`.
 
   The query is compiled and runs efficiently against the dataframe.
-  The query must return a boolean expression. Given this function is
-  a macro, you must require `Explorer.DataFrame` before using it.
+  The query must return a boolean expression.
+
+  > #### Notice {: .notice}
+  >
+  > Given this function is a macro, you must require `Explorer.DataFrame` before using it.
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:1334]:
```suggestion
  > This is macro, therefore you must `require  Explorer.DataFrame` before using it.
```

File: lib/explorer/data_frame.ex [hunk 2]
```diff
@@ -1478,15 +1479,23 @@ defmodule Explorer.DataFrame do
...
-  Creates and modifies columns.
+  Creates or modifies columns based on `Explorer.Query`.
+
+  The query is compiled and runs efficiently against the dataframe.
+  New variables overwrite existing variables of the same name.
+  Column names are coerced from atoms to strings.
+
+  > #### Notice {: .notice}
+  >
+  > Given this function is a macro, you must require `Explorer.DataFrame` before using it.
```
Comment (josevalim) [inline lib/explorer/data_frame.ex:1490]:
```suggestion
  > This is macro, therefore you must `require  Explorer.DataFrame` before using it.
```

Comment (josevalim) [reply lib/explorer/data_frame.ex:1490]: Another idea is to find all places where we do  `alias Explorer.DataFrame, as: DF` and replace it by `require Explorer.DataFrame, as: DF`.

**Review by josevalim — APPROVED** [review]
Beautiful!

File: lib/explorer/data_frame.ex (hunk 2 above)
Comment (philss) [reply lib/explorer/data_frame.ex:1490]: I'm going to try that in the next PR. Thanks! :)

**Pull request merged**